src/ucp/core/ucp_am.c

```diff
@@ -1394,50 +1443,74 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_am_long_first_handler,
     /* This is the first fragment, other fragments (if arrived) should be on
      * ep_ext->am.mid_rdesc_q queue */
     ucs_assert(NULL == ucp_am_find_first_rdesc(worker, ep_ext,
-                                               first_hdr->msg_id));
+                                               first_ftr->super.msg_id));
 
     /* Alloc buffer for the data and its desc, as we know total_size.
      * Need to allocate a separate rdesc which would be in one contigious chunk
-     * with data buffer. */
-    first_rdesc = ucs_malloc(total_length + sizeof(ucp_recv_desc_t),
+     * with data buffer. The layout of assembled message is below:
+     *
+     * +-------+-----------+--------+---------+---------+----------+
+     * | rdesc | first_ftr | am_hdr | padding | payload | user hdr |
```

Comment:
maybe simpler if first_ftr is after user_hdr, as in the wire format?

Reply:
we have to access `first_ftr` while processing the fragments, this way it's simpler to get an access to it

Reply:
ok, many note this in comment

Reply:
ok (i guess you meant to add this description to the comment)